test/models/papi_client_test.rb

```diff
@@ -0,0 +1,254 @@
+      converted_name = @client.to_cli_opt(opt_name)
+      assert de_cmd.include? converted_name
+      assert de_cmd.include? opt_val
+    end
+    assert de_cmd.include? '--differential-expression'
+
+    # test sparse matrix args
+    de_sparse_opts = {
+      annotation_name: 'Category',
+      annotation_type: 'group',
+      annotation_scope: 'cluster',
+      annotation_file: @cluster_file.gs_url,
+      cluster_file: @cluster_file.gs_url,
+      cluster_name: 'cluster.txt',
+      matrix_file_path: 'gs://test_bucket/expression/sparse.mtx',
+      matrix_file_type: 'sparse',
+      gene_file: 'gs://test_bucket/expression/genes.tsv',
+      barcode_file: 'gs://test_bucket/expression/barcodes.tsv'
+    }
+
+    de_sparse_cmd = @client.get_de_parameters(de_sparse_opts)
+    assert de_sparse_cmd.any?
+    de_sparse_opts.each do |opt_name, opt_val|
+      converted_name = @client.to_cli_opt(opt_name)
+      assert de_sparse_cmd.include? converted_name
+      assert de_sparse_cmd.include? opt_val
+    end
+    assert de_sparse_cmd.include? '--differential-expression'
+  end
+
+  test 'should get extra command line options' do
+    cluster_cmd = @client.get_command_line(study_file: @cluster_file, action: :ingest_cluster,
+                                           user_metrics_uuid: @user.metrics_uuid)
+    assert cluster_cmd.include? '--domain-ranges'
+    sanitized_domains = "{'x':[-1,1],'y':[-2,2],'z':[-3,3]}"
+    assert cluster_cmd.include? sanitized_domains
+  end
+
+  # this test covers many sub-methods that are required to create a pipeline request, such as creating resources,
+  # environment, actions, and pipelines objects
+  test 'should create pipelines request object' do
+    exp_cmd = @client.get_command_line(study_file: @expression_matrix, action: :ingest_expression,
+                                       user_metrics_uuid: @user.metrics_uuid)
+    environment = @client.set_environment_variables
+    actions = @client.create_actions_object(commands: exp_cmd, environment: environment)
+    regions = %w[us-central1]
+    resources = @client.create_resources_object(regions: regions)
+    pipeline = @client.create_pipeline_object(actions: actions, environment: environment, resources: resources)
+    labels = { foo: 'bar' }
+    pipeline_request = @client.create_run_pipeline_request_object(pipeline: pipeline, labels: labels)
+    assert pipeline_request.is_a? Google::Apis::GenomicsV2alpha1::RunPipelineRequest
+    assert_equal pipeline, pipeline_request.pipeline
+    assert_equal actions, pipeline_request.pipeline.actions
+    assert_equal environment, pipeline_request.pipeline.environment
+    assert_equal resources, pipeline_request.pipeline.resources
+    assert_equal exp_cmd, pipeline_request.pipeline.actions.commands
+  end
+
+  test 'converts keys into cli options' do
+    assert_equal '--annotation-name', @client.to_cli_opt(:annotation_name)
+    assert_equal '--foo', @client.to_cli_opt('foo')
+  end
```

Comment:
These are all extremely readable tests, which is impressive given the somewhat arcane nature of what the code accomplishes.  Nice work!